malicious-site-protection/malicious-site-protection-impl/src/main/kotlin/com/duckduckgo/malicioussiteprotection/impl/data/MaliciousSiteRepository.kt

```diff
@@ -0,0 +1,148 @@
+    suspend fun matches(hashPrefix: String): List<Match>
+    suspend fun loadFilters()
+    suspend fun loadHashPrefixes()
+}
+
+@ContributesBinding(AppScope::class)
+@SingleInstanceIn(AppScope::class)
+class RealMaliciousSiteRepository @Inject constructor(
+    private val maliciousSiteDao: MaliciousSiteDao,
+    private val maliciousSiteService: MaliciousSiteService,
+) : MaliciousSiteRepository {
+
+    override suspend fun loadFilters() {
+        try {
+            val networkRevision = maliciousSiteService.getRevision().revision
+
+            (maliciousSiteDao.getLatestRevision()?.filter { it.type == Type.FILTER_SET.name } ?: listOf()).let { latestRevision ->
+                val phishingFilterSetRevision = latestRevision.firstOrNull() { it.feed == PHISHING.name }?.revision ?: 0
+                val phishingFilterSet: FilterSetResponse? = if (networkRevision > phishingFilterSetRevision) {
+                    maliciousSiteService.getPhishingFilterSet(phishingFilterSetRevision)
+                } else {
+                    null
+                }
+                val malwareFilterSetRevision = latestRevision.firstOrNull() { it.feed == MALWARE.name }?.revision ?: 0
+                val malwareFilterSet: FilterSetResponse? = if (networkRevision > malwareFilterSetRevision) {
+                    maliciousSiteService.getMalwareFilterSet(malwareFilterSetRevision)
+                } else {
+                    null
+                }
+
+                maliciousSiteDao.updateFilters(
+                    phishingFilterSet?.let {
+                        PhishingFilterSetWithRevision(
+                            it.insert.map { insert -> Filter(insert.hash, insert.regex) }.toSet(),
+                            it.delete.map { delete -> Filter(delete.hash, delete.regex) }.toSet(),
+                            it.revision,
+                            it.replace,
+                        )
+                    },
+                )
+                maliciousSiteDao.updateFilters(
+                    malwareFilterSet?.let {
+                        MalwareFilterSetWithRevision(
+                            it.insert.map { insert -> Filter(insert.hash, insert.regex) }.toSet(),
+                            it.delete.map { delete -> Filter(delete.hash, delete.regex) }.toSet(),
+                            it.revision,
+                            it.replace,
+                        )
+                    },
+                )
+            }
+        } catch (e: Exception) {
+            Timber.e(e, "Failed to download malicious site protection list")
+        }
+    }
+
+    override suspend fun loadHashPrefixes() {
+        try {
+            val networkRevision = maliciousSiteService.getRevision().revision
+
+            (maliciousSiteDao.getLatestRevision()?.filter { it.type == Type.HASH_PREFIXES.name } ?: listOf()).let {
+                val phishingHashPrefixesRevision = it.firstOrNull() { it.feed == PHISHING.name }?.revision ?: 0
```

Comment:
NIT: there's some repeated code 4 times in this class, we can consider creating a more generic logic that can be reused.